### PR TITLE
Feature/dropped monitoring frames

### PR DESCRIPTION
**Pull request by hslusarek**
This PR adds functionality which warns the user if monitoring frames are missing or if no monitoring frames are received.

### CODE REVIEW

**Review by martiniil — CHANGES_REQUESTED** [review]

File: include/psen_scan_v2/complete_scan_validator.h [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    ScanRoundInfo(const uint32_t& scan_counter);
+
+  public:
+    Result validate(const uint32_t expected_num_msgs) const;
+    uint32_t scanCounter() const;
+    ScanRoundInfo& operator++();
+
+  private:
+    uint32_t scan_counter_{ 0 };
+    unsigned short int num_received_msgs_{ 1 };
```
[the remark this answers is not in the document]
Comment (hslusarek) [reply include/psen_scan_v2/complete_scan_validator.h:81]: Changed -> Done.

File: include/psen_scan_v2/scanner_state_machine.h [hunk 2]
```diff
@@ -64,22 +66,36 @@ namespace e = psen_scan_v2::scanner_protocol::scanner_events;
 }
 // clang-format on
 
+static constexpr std::chrono::milliseconds WATCHDOG_TIMEOUT{ 1000 };
```
[the remark this answers is not in the document]
Comment (hslusarek) [reply include/psen_scan_v2/scanner_state_machine.h:69]: Done.

File: include/psen_scan_v2/scanner_state_machine.hpp [hunk 3]
```diff
@@ -23,23 +23,22 @@ inline ScannerProtocolDef::ScannerProtocolDef(StateMachineArgs* const args) : ar
...
+  PSENSCAN_DEBUG("StateMachine", fmt::format("Entering state: {}", #state_name));\
   }\
 
 #define DEFAULT_ON_EXIT_IMPL(state_name)\
   template <class Event, class FSM>\
   void ScannerProtocolDef::state_name::on_exit(Event const&, FSM& fsm)\
-  {\
-    PSENSCAN_DEBUG("StateMachine", fmt::format("Exiting state: {}", #state_name));\
+{\
+  PSENSCAN_DEBUG("StateMachine", fmt::format("Exiting state: {}", #state_name));\
```
[the remark this answers is not in the document]
Comment (hslusarek) [reply include/psen_scan_v2/scanner_state_machine.hpp:34]: No, this was not intended. -> Done.

File: include/psen_scan_v2/scanner_state_machine.hpp [hunk 4]
```diff
@@ -52,14 +51,39 @@ void ScannerProtocolDef::Idle::on_exit(Event const&, FSM& fsm)
...
   fsm.args_->scanner_started_cb();
 }
 
-DEFAULT_ON_EXIT_IMPL(WaitForMonitoringFrame)
+template <class Event, class FSM>
+void ScannerProtocolDef::WaitForMonitoringFrame::on_exit(Event const&, FSM& fsm)
+{
+  PSENSCAN_DEBUG("StateMachine", fmt::format("Exiting state: {}", "WaitForMonitoringFrame"));
+  // Stop watchdog...
+  fsm.monitoring_frame_watchdog_.reset();
```
[the remark this answers is not in the document]
Comment (hslusarek) [reply include/psen_scan_v2/scanner_state_machine.hpp:85]: I added a `reset()` method to the `Watchdog` class which is now used to restart the `Watchdog` timer whenever a new `MonitoringFrame` is received.

File: test/integration_tests/integrationtest_scanner_api.cpp [hunk 5]
```diff
@@ -64,13 +67,35 @@ using namespace ::testing;
...
+  static std::default_random_engine re{};
+  using Dist = std::uniform_real_distribution<double>;
+  static Dist uid{};
+  return uid(re, Dist::param_type{ low, high });
+}
+
+static std::vector<double> createRandomVec(const unsigned int& num_elements, const double& low, const double& high)
+{
+  std::vector<double> vec(num_elements);
+  std::generate(vec.begin(), vec.end(), [low, high]() { return std::round(randDouble(low, high)); });
```
[the remark this answers is not in the document]
Comment (hslusarek) [reply test/integration_tests/integrationtest_scanner_api.cpp]: Done.

File: test/integration_tests/integrationtest_scanner_api.cpp [hunk 6]
```diff
@@ -64,13 +67,41 @@ using namespace ::testing;
...
+
+static double restrictToOneDigitsAfterComma(const double& value)
+{
+  return std::round(value * 10.) / 10.;
+}
+
+static std::vector<double> createRandomVec(const unsigned int& num_elements, const double& low, const double& high)
+{
+  std::vector<double> vec(num_elements);
+  // The scanner sends tenth degree values. Therefore, restrict values to one digit after the comma.
```
Comment (martiniil) [inline test/integration_tests/integrationtest_scanner_api.cpp:86]: This is correct for the measurements, but not for the intensities. There you need integer values.

**Review by martiniil — CHANGES_REQUESTED** [review]
Only minor things this time. Feel free to merge once they are adressed.

File: include/psen_scan_v2/watchdog.h [hunk 7]
```diff
@@ -16,39 +16,59 @@
 #define PSEN_SCAN_V2_WATCHDOG_H
 
 #include <chrono>
-#include <functional>
```
[the remark this answers is not in the document]
Comment (hslusarek) [reply include/psen_scan_v2/watchdog.h:19]: Done.

File: test/integration_tests/integrationtest_scanner_api.cpp (hunk 6 above)
Comment (hslusarek) [reply test/integration_tests/integrationtest_scanner_api.cpp:86]: Done.

**Pull request merged**